### PR TITLE
Expose DWARFDIE::GetDeclContext() in lldb_private::Function.

### PR DESCRIPTION
**Pull request by adrian-prantl**
I need this API in the Swift plugin, but it seems generally useful enough to expose it in the main branch.

### CODE REVIEW

File: lldb/include/lldb/Symbol/Function.h [hunk 1]
```diff
@@ -533,6 +533,12 @@ class Function : public UserID, public SymbolContextScope {
   ///     The DeclContext, or NULL if none exists.
   CompilerDeclContext GetDeclContext();
 
+  /// Get the CompilerContext for this function, if available.
+  ///
+  /// \return
+  ///     The CompilerContext, or NULL if none exists.
+  std::vector<CompilerContext> GetCompilerContext();
```
Comment (bulbazord) [inline lldb/include/lldb/Symbol/Function.h:540]: The docs and the declaration conflict.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 2]
```diff
@@ -142,8 +142,7 @@ TypeSP DWARFASTParserClang::ParseTypeFromClangModule(const SymbolContext &sc,
   // If this type comes from a Clang module, recursively look in the
   // DWARF section of the .pcm file in the module cache. Clang
   // generates DWO skeleton units as breadcrumbs to find them.
-  llvm::SmallVector<CompilerContext, 4> decl_context;
-  die.GetDeclContext(decl_context);
+  std::vector<CompilerContext> decl_context = die.GetDeclContext();
```
Comment (bulbazord) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:145]: Why the change from `llvm::SmallVector<$Type, 4>` to `std::vector`? Does this not generally contain 4 or fewer CompilerContexts anymore?

**Review by walter-erquinigo — APPROVED** [review]

File: lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARFDebugMap.cpp [hunk 3]
```diff
@@ -1391,7 +1391,16 @@ SymbolFileDWARFDebugMap::GetDeclContextContainingUID(lldb::user_id_t type_uid) {
...
     return oso_dwarf->GetDeclContextContainingUID(type_uid);
-  return CompilerDeclContext();
+  return {};
+}
+
+std::vector<CompilerContext>
+SymbolFileDWARFDebugMap::GetCompilerContextForUID(lldb::user_id_t type_uid) {
+  const uint64_t oso_idx = GetOSOIndexFromUserID(type_uid);
+  SymbolFileDWARF *oso_dwarf = GetSymbolFileByOSOIndex(oso_idx);
+  if (oso_dwarf)
```
Comment (bulbazord) [inline lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARFDebugMap.cpp:1401]: You can roll these lines into 1
```
if (SymbolFileDWARF *oso_dwarf = GetSymbolFileByOSOIndex(oso_idx))
```

File: lldb/source/Symbol/Function.cpp [hunk 4]
```diff
@@ -519,7 +530,17 @@ CompilerDeclContext Function::GetDeclContext() {
...
       return sym_file->GetDeclContextForUID(GetID());
   }
-  return CompilerDeclContext();
+  return {};
+}
+
+std::vector<CompilerContext> Function::GetCompilerContext() {
+  ModuleSP module_sp = CalculateSymbolContextModule();
+
+  if (module_sp) {
```
Comment (bulbazord) [inline lldb/source/Symbol/Function.cpp:539]: You can roll these into 1 line as well.

File: lldb/include/lldb/Symbol/Function.h (hunk 1 above)
Comment (JDevlieghere) [inline lldb/include/lldb/Symbol/Function.h:540]: This comment is wrong/out of date.

File: lldb/tools/lldb-test/lldb-test.cpp [hunk 5]
```diff
@@ -816,6 +818,8 @@ Expected<Error (*)(lldb_private::Module &)> opts::symbols::getAction() {
           "Specify search type (-find) to use search options.");
     return dumpModule;
 
+
+
```
Comment (bulbazord) [inline lldb/tools/lldb-test/lldb-test.cpp:822]: Spurious whitespace

**Review by bulbazord — COMMENTED** [review]
overall LGTM, few questions though

File: lldb/source/Symbol/Function.cpp [hunk 6]
```diff
@@ -396,6 +396,17 @@ void Function::GetDescription(Stream *s, lldb::DescriptionLevel level,
     s->AsRawOstream() << ", name = \"" << name << '"';
   if (mangled)
     s->AsRawOstream() << ", mangled = \"" << mangled << '"';
+  if (level == eDescriptionLevelVerbose) {
+    *s << ", decl_context = {";
+    auto decl_context = GetCompilerContext();
+    // Drop the function itself from the context chain.
+    if (decl_context.size())
+      decl_context.pop_back();
+    llvm::interleave(
```
Comment (Michael137) [inline lldb/source/Symbol/Function.cpp:405]: can't you use `interleaveComma` here?

File: lldb/tools/lldb-test/lldb-test.cpp [hunk 7]
```diff
@@ -322,10 +322,12 @@ std::vector<CompilerContext> parseCompilerContext() {
     }
     result.push_back({kind, ConstString{value}});
   }
-  outs() << "Search context: {\n";
-  for (auto entry: result)
-    entry.Dump();
-  outs() << "}\n";
+  outs() << "Search context: {";
+  lldb_private::StreamString s;
+  llvm::interleave(
```
Comment (Michael137) [inline lldb/tools/lldb-test/lldb-test.cpp:327]: `interleaveComma`

File: lldb/source/Symbol/Function.cpp (hunk 6 above)
Comment (Michael137) [reply lldb/source/Symbol/Function.cpp:405]: Oh unless it doesn't work with this particular stream...

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp (hunk 2 above)
Comment (adrian-prantl) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:145]: This was purely an aesthetic choice. I felt like hardcoding the constant 4 makes a choice for something that's hugely language-dependent in a general API, and given that this is parsing DWARF the extra malloc overhead of std::vector likely does not matter in the slightest.

File: lldb/source/Symbol/Function.cpp (hunk 6 above)
Comment (adrian-prantl) [reply lldb/source/Symbol/Function.cpp]: No, it's templated!

File: lldb/tools/lldb-test/lldb-test.cpp (hunk 7 above)
Comment (adrian-prantl) [reply lldb/tools/lldb-test/lldb-test.cpp]: Yes, that is even shorter!

File: lldb/test/Shell/SymbolFile/DWARF/x86/module-ownership.mm [hunk 8]
```diff
@@ -26,7 +26,7 @@
 // CHECK-ANON-S1: CXXRecordDecl {{.*}} imported in A struct
 
 StructB s3;
-// CHECK-ANON-S2: CXXRecordDecl {{.*}} imported in A.B {{.*}} struct
+// CHECK-ANON-S2: CXXRecordDecl {{.*}} imported in A.B {{.*}}struct
```
Comment (adrian-prantl) [inline lldb/test/Shell/SymbolFile/DWARF/x86/module-ownership.mm:29]: In case someone asks about this: before we force-parsed the DIE by calling GetDeclContext() there used to be an `<undeserialized declaration>` output here that is irrelevant to the test.

File: lldb/include/lldb/Symbol/Type.h [hunk 9]
```diff
@@ -34,7 +34,7 @@ struct CompilerContext {
   }
   bool operator!=(const CompilerContext &rhs) const { return !(*this == rhs); }
 
-  void Dump() const;
+  void Dump(Stream *s) const;
```
Comment (clayborg) [inline lldb/include/lldb/Symbol/Type.h:37]: I think Alex Langford is trying to get rid of functions that take "Stream *s" and replace them with "Stream &s"

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFDIE.cpp [hunk 10]
```diff
@@ -373,47 +373,49 @@ std::vector<DWARFDIE> DWARFDIE::GetDeclContextDIEs() const {
...
     break;
   case DW_TAG_subprogram:
-    context.push_back(
-        {CompilerContextKind::Function, ConstString(GetPubname())});
+    push_ctx(CompilerContextKind::Function, GetPubname());
     break;
   case DW_TAG_variable:
-    context.push_back(
-        {CompilerContextKind::Variable, ConstString(GetPubname())});
+push_ctx(CompilerContextKind::Variable, GetPubname());
```
Comment (clayborg) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFDIE.cpp:410]: indent

File: lldb/source/Symbol/Type.cpp [hunk 11]
```diff
@@ -64,49 +64,51 @@ bool lldb_private::contextMatches(llvm::ArrayRef<CompilerContext> context_chain,
   return true;
 }
 
-void CompilerContext::Dump() const {
+void CompilerContext::Dump(Stream *s) const {
```
Comment (clayborg) [inline lldb/source/Symbol/Type.cpp:67]: "Stream &s"

File: lldb/include/lldb/Symbol/Type.h (hunk 9 above)
Comment (bulbazord) [reply lldb/include/lldb/Symbol/Type.h:37]: Yes, we basically always want the stream to be a valid pointer, a reference makes more sense here.

Comment (adrian-prantl) [reply lldb/include/lldb/Symbol/Type.h:37]: Sounds good to me!
